extensions/binding/org.eclipse.smarthome.binding.mqttbroker/src/main/java/org/eclipse/smarthome/binding/mqttbroker/internal/ssl/Pin.java

```diff
@@ -0,0 +1,106 @@
+package org.eclipse.smarthome.binding.mqttbroker.internal.ssl;
+
+import java.util.Arrays;
+
+import org.eclipse.smarthome.binding.mqttbroker.internal.HexUtils;
+
+/**
+ * A pin is either a public key pin or certificate pin and consists of the binary data
+ * and the used hash algorithm.
+ *
+ * @author David Graeff - Initial contribution
+ */
+public class Pin {
+    protected byte[] pinData;
+    protected HashMethod method;
+    protected boolean learning;
+    protected final PinType type;
+
+    /**
+     * To simplify the creating of a Pin instance, you can use the factory
+     * methods `newCertificatePin` and `newPublicKeyPin` of this class.
+     *
+     * @param type The pin type
+     * @param method The hash method
+     * @param learning If the Pin is in learning mode.
+     * @param pinData The pinned data
+     */
+    Pin(PinType type, HashMethod method, boolean learning, byte[] pinData) {
+        this.type = type;
+        this.method = method;
+        this.learning = learning;
+        this.pinData = pinData;
+    }
+
+    public PinType getType() {
+        return type;
+    }
+
+    public byte[] getHash() {
+        return pinData;
+    }
+
+    public void setLearningMode() {
+        this.learning = true;
+        this.pinData = null;
+    }
+
+    /**
+     * This sets the pin instance to checking mode. The given
+     * data will be used, it is expected to be a SHA256 hash.
+     *
+     * @param data SHA256 hash data
+     */
+    public void setCheckMode(byte[] data) {
+        this.learning = false;
+        this.pinData = data;
+    }
+
+    public static Pin newCertificatePin(HashMethod method, boolean learning, byte[] pinData) {
+        if (!learning && pinData == null) {
+            throw new IllegalArgumentException("data cannot be null if pinning method is set.");
+        }
+        return new Pin(PinType.CERTIFICATE_TYPE, method, learning, pinData);
+    }
+
+    public static Pin newPublicKeyPin(HashMethod method, boolean learning, byte[] pinData) {
+        if (!learning && pinData == null) {
+            throw new IllegalArgumentException("data cannot be null if pinning method is set.");
+        }
+        return new Pin(PinType.PUBLIC_KEY_TYPE, method, learning, pinData);
+    }
+
+    /**
+     * Returns true if this pin is still learning.
+     */
+    public boolean isLearning() {
+        return learning;
+    }
+
+    /**
+     * This method is used to determine if the given digest is equal to the
+     * one of this Pin. If this Pin is still learning, it will always return true.
+     *
+     * @param digestData SHA256 hash data of a public key or a certificate.
+     * @return Returns true if equal
+     */
+    public boolean isEqual(byte[] digestData) {
+        if (learning) {
+            return true;
+        }
+        return Arrays.equals(pinData, digestData);
+    }
+
+    @Override
+    public String toString() {
+        return method.name() + ":" + HexUtils.toHexString(pinData);
```

Comment:
This gives a NPE if pinData is null (which seems to be a valid situation).

Reply:
You might want to add some null annotations throughout your helper classes/methods to avoid such situations.